lib/expansion_rules.rb

```diff
@@ -148,22 +148,41 @@ def dependency_resolution
     @dependency_resolution ||= ExpansionRules::DependencyResolution.new(self)
   end
 
-  def find_custom_expansion_fields(document_type, options = {})
-    should_check_link_type = options[:link_type]
-    link_type = options[:link_type].try(:to_sym)
+  def expansion_fields_for_document_type(document_type)
+    matching_document_types = CUSTOM_EXPANSION_FIELDS.select do |item|
+      item[:document_type] == document_type.to_sym
+    end
+
+    return DEFAULT_FIELDS unless matching_document_types.any?
 
-    condition = CUSTOM_EXPANSION_FIELDS.find do |cond|
-      next if should_check_link_type && cond.fetch(:link_type, link_type) != link_type
+    collated_fields = matching_document_types.inject([]) { |memo, item| memo + item[:fields] }
+    may_use_default_fields = matching_document_types.all? { |item| item[:link_type].present? }
+
+    collated_fields += DEFAULT_FIELDS if may_use_default_fields
+    collated_fields.uniq
+  end
 
-      cond[:document_type] == document_type.to_sym
+  def expansion_fields_for_linked_document_type(document_type, link_type)
+    matching_link = CUSTOM_EXPANSION_FIELDS.find do |item|
+      item[:document_type] == document_type.to_sym &&
+        item[:link_type] == link_type.to_sym
     end
+    return matching_link[:fields] if matching_link
 
-    condition[:fields] if condition
+    matching_document_type = CUSTOM_EXPANSION_FIELDS.find do |item|
+      item[:document_type] == document_type.to_sym && item[:link_type].nil?
+    end
+    return matching_document_type[:fields] if matching_document_type
+
+    DEFAULT_FIELDS
   end
 
   def expansion_fields(document_type, link_type = nil)
-    find_custom_expansion_fields(document_type, link_type: link_type) ||
-      DEFAULT_FIELDS
+    if link_type
+      expansion_fields_for_linked_document_type(document_type, link_type)
+    else
+      expansion_fields_for_document_type(document_type)
+    end
```

Comment:
I would move this method up to be above the definitions for `expansion_fields_for_linked_document_type` and `expansion_fields_for_document_type`, so that it reads top-to-bottom. Same with the accompanying tests.

Reply:
Ok done.